radixdlt/src/integration/java/com/radixdlt/consensus/simulation/checks/SafetyCheck.java

```diff
@@ -17,34 +17,105 @@
+				String.format("Different vertices [%s, %s] committed at same view: %s",
+					vertex,
+					currentVertex,
+					vertex.getView()
+				)
+			)
+		);
+	}
+
+	private static Observable<BFTCheckError> noParentError(Vertex vertex) {
+		return Observable.just(
+			new BFTCheckError(
+				String.format("Committed vertex %s has no parent", vertex)
+			)
+		);
+	}
+
+	private static Observable<BFTCheckError> badParentError(Vertex vertex, Vertex lastCommitted) {
+		return Observable.just(
+			new BFTCheckError(
+				String.format("Parent of vertex %s doesn't match last committed %s",
+					vertex,
+					lastCommitted
+				)
+			)
+		);
+	}
+
 	@Override
 	public Observable<BFTCheckError> check(SimulatedBFTNetwork network) {
-		return Observable.zip(
-			network.getNodes().stream()
-				.map(network::getVertexStore)
-				.map(VertexStore::lastCommittedVertex)
-				.collect(Collectors.toList()),
-				Arrays::stream
+		final Map<View, Vertex> committedVertices = new ConcurrentHashMap<>();
+		committedVertices.put(View.genesis(), network.getGenesisVertex());
+		final AtomicReference<View> highest = new AtomicReference<>(View.genesis());
+		final Map<ECKeyPair, Vertex> lastCommittedByNode = new ConcurrentHashMap<>();
+
+		return Observable.merge(
+			network.getNodes().stream().map(
+				node -> network.getVertexStore(node)
+					.lastCommittedVertex().map(v -> Pair.of(node, v)))
+				.collect(Collectors.toList())
 		)
-			.map(committedVertices -> committedVertices.distinct().collect(Collectors.toList()))
-			.concatMap(committedVertices -> {
-				if (committedVertices.size() != 1) {
-					return Observable.just(new BFTCheckError("Committed vertices don't match: " + committedVertices));
+			.flatMap(nodeAndVertex -> {
+				final ECKeyPair node = nodeAndVertex.getFirst();
+				final Vertex vertex = nodeAndVertex.getSecond();
+				final Vertex currentVertexAtView = committedVertices.get(vertex.getView());
+				if (currentVertexAtView != null) {
+					if (!currentVertexAtView.getId().equals(vertex.getId())) {
+						return differentVerticesError(vertex, currentVertexAtView);
+					}
 				} else {
-					return Observable.empty();
+					final Vertex lastCommitted = committedVertices.get(highest.get());
+					if (vertex.getParentId() == null) {
+						return noParentError(vertex);
+					}
+
+					if (!vertex.getParentId().equals(lastCommitted.getId())
+						|| !vertex.getParentView().equals(lastCommitted.getView())) {
+						return badParentError(vertex, lastCommitted);
+					}
+
+					committedVertices.put(vertex.getView(), vertex);
+					highest.set(vertex.getView());
 				}
+
+				// Clean up old vertices so that we avoid consuming too much memory
+				lastCommittedByNode.put(node, vertex);
+				final View lowest = lastCommittedByNode.values().stream()
+					.map(Vertex::getView)
+					.reduce((v0, v1) -> v0.compareTo(v1) < 0 ? v0 : v1)
+					.orElse(View.genesis());
+				final Set<View> viewsToRemove = committedVertices.keySet().stream()
```

Comment:
Skip using sets and use array instead to get deterministic behaviour? Or is it important to have non-determinism here? :) For uniqueness you can use `.distinct()` I think?

Reply:
Not sure what that gains since deterministic iteration order doesn't matter in this logic.
Also, there's already a lot of "non-determinism" in here like iterating through keys of the map or `.reduce` function above.